### PR TITLE
Improved Compiler API 1

### PR DESCRIPTION
**Pull request by alex-mckenna**
A first move towards the goals of #1932. This PR introduces the `ClashEnv` and `ClashDesign` types, and makes some basic changes to tidy up the public interfaces of functions like `generateBindings`, `generateHDL` and `initBackend`. See the individual commit messages for more information.

This PR should be prioritised for 1.6, merging post-release will _ruin_ backports for the 1.6 series of releases, which I very much want to avoid. Further work to #1932 will be delayed until 1.8 is planned (unless it can be implemented without making backports more difficult).

## Still TODO:

  - [ ] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

[comment]: # (Depending on the change, some of these points may not be necessary. If so, leave the boxes unchecked so it is easier for a reviewer to see what has been omitted.)

### CODE REVIEW

File: clash-lib/tests/Test/Clash/Rewrite.hs [hunk 1]
```diff
@@ -60,16 +60,28 @@ lookupTM u tm = case HashMap.lookup u tm of
...
+            , opt_aggressiveXOpt = False
+            , opt_inlineLimit = 20
+            , opt_inlineFunctionLimit = 15
+            , opt_inlineConstantLimit = 0
+            , opt_inlineWFCacheLimit = 10
+            , opt_specLimit = 20
+            , opt_evaluatorFuelLimit = 10
+            , opt_newInlineStrat = True
+            , opt_ultra = False
+            }
```
Comment (martijnbastiaan) [inline clash-lib/tests/Test/Clash/Rewrite.hs:75]: Why aren't these part of `defClashOpts`?

File: Clash.hs [hunk 2]
```diff
@@ -17,55 +14,57 @@ import Clash.Backend
...
-import Util (OverridingBool(..))
-#endif
-
 genSystemVerilog
-  :: String
+  :: ClashOpts
+  -> String
   -> IO ()
-genSystemVerilog = doHDL (initBackend WORD_SIZE_IN_BITS HDLSYN True PreserveCase Nothing (AggressiveXOptBB False) (RenderEnums True) :: SystemVerilogState)
+genSystemVerilog = doHDL (Proxy @SystemVerilogState)
```
Comment (martijnbastiaan) [inline Clash.hs:28]: This has annoyed me for sooooo long :star2:

File: clash-lib/src/Clash/Driver/Types.hs [hunk 3]
```diff
@@ -362,6 +390,35 @@ data ClashOpts = ClashOpts
...
+  rnf o =
+    opt_inlineLimit o `deepseq`
+    opt_specLimit o `deepseq`
+    opt_inlineFunctionLimit o `deepseq`
+    opt_inlineConstantLimit o `deepseq`
+    opt_evaluatorFuelLimit o `deepseq`
+    opt_cachehdl o `deepseq`
+    opt_clear o `deepseq`
+    opt_primWarn o `deepseq`
+    opt_color o `seq`
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Driver/Types.hs:403]: We should probably add our own type with NFData..

**Review by martijnbastiaan — APPROVED** [review]
A clear improvement over the status quo!

File: clash-lib/tests/Test/Clash/Rewrite.hs (hunk 1 above)
Comment (alex-mckenna) [reply clash-lib/tests/Test/Clash/Rewrite.hs:75]: Not sure, I didn't double check _how_ they differed from `defClashOpts`. I wouldn't expect anything in this test to need limits raising...

Comment (alex-mckenna) [reply clash-lib/tests/Test/Clash/Rewrite.hs]: The only meaningful change was `opt_debug` so I removed the others

File: clash-lib/src/Clash/Driver/Types.hs (hunk 3 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Driver/Types.hs:403]: Yeah... there are some other GHC types we use directly which I think are better replaced with analogues in `clash-lib`, like `InlineSpec` and `SrcLoc`. Even if they're the same, using them directly from `ghc` is a weird coupling between the front-end and the back-end of the compiler IMO

Comment (alex-mckenna) [reply clash-lib/src/Clash/Driver/Types.hs:403]: But it's a bit tangential, I'll save it for another PR. I think I'll spend the first few days next week doing some more _get ready for 1.6_ PRs

**Pull request merged**